File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        //
        // O(1) SPACE LOGIC EXPLAINED -- Suman Barick
        //
        // There is an interesting observation.
        // If we write down all the elements of List A
        // and then all the elements of list B
        // for this merged list (A + B)
        // if the intersection node appears at index x,
        // It will be the same for (B + A)

        // So, by this logic, we will have 2 pointers,
        // p1 will traverse through (A + B)
        // and p2 will traverse through
        ListNode p1 = headA;
        ListNode p2 = headB;

        Boolean p1Switched = false;
        Boolean p2Switched = false;

        // since (A+B) is of same length of (B+A)
        // does not matter which pointer we check for termination
        while (p1 != null && p2 != null) {
            // match/intersection found? break
            if(p1 == p2) {
                break;
            }
            // else move on
            p1 = p1.next;
            p2 = p2.next;

            // both of them cannot be null at the SAME time if they have intersection
            // end reached for any of them?
            // start with the next 
            if(p1 == null && !p1Switched) {
                p1 = headB;
                p1Switched = true;
            }
            if(p2 == null && !p2Switched) {
                p2 = headA;
                p2Switched = true;
            }
        }

        return p1;
    }
}

File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function(headA, headB) {
    // Method 1:
    // if they have an intersection, then they will meet 
    // somewhere <= before tail
    // So, if we calculate the length of both, 
    // And then skip the first (bigLength - smallLength) no of node
    // from the bigger LL, and then start running 1 pointer for each
    // both the pointers should run side by side
    let l1 = getLen(headA);
    let l2 = getLen(headB);

    let diffL = Math.abs(l1-l2);
    let i = headA, j = headB;

    // fast forward 1 pointer so that the remain lenth remain same for both
    for(let k=1; k <= diffL; k++) {
        l1 > l2 ? i = i.next : j = j.next;
    }

    // now keep moving them side by side
    // and check at any point if they are pointing to the same node
    // that will be the intersecting node
    while(i && j) {
        if(i === j) return i;
        i = i.next;
        j = j.next;
    }

    return null;
};

function getLen(head) {
    let node = head;
    let l = 0;
    while(node) {
        l++;
        node = node.next;
    }

    return l;
}